File: apps/bots/src/parser/modules/faceit.ts
import { FaceitDbData } from '../../integrations/faceit.js';
import { Module, State } from '../index.js';

const getFaceitIntegrationData = async (state: State) => {
  const integrations = await state.cache.getEnabledIntegrations();
  const integration = integrations.find((i) => i.integration.service === 'FACEIT');

  if (!integration) return 'Integration with faceit not enabled';
  const integrationData = integration.data as unknown as FaceitDbData;

  const data = await state.cache.getFaceitData(integrationData.username, integrationData.game);

  return data;
};

export const faceit: Module[] = [
  {
    key: 'faceit.elo',
    description: 'Faceit elo',
    handler: async (_key, state) => {
      const data = await getFaceitIntegrationData(state);
      if (!data) {
        return 'Cannot fetch faceit data.';
      }
      if (typeof data === 'string') {
        return data;
      }

      return data.elo;
    },
  },
  {
    key: 'faceit.lvl',
    description: 'Faceit Lvl',
    handler: async (_key, state) => {
      const data = await getFaceitIntegrationData(state);
      if (!data) {
        return 'Cannot fetch faceit data.';
      }
      if (typeof data === 'string') {
        return data;
      }

      return data.lvl;
    },
  },
  {
    key: 'faceit.todayEloDiff',
    description: 'Faceit today elo earned',
    handler: async (_key, state) => {
      const data = await getFaceitIntegrationData(state);
      if (!data) {
        return 'Cannot fetch faceit data.';
      }
      if (typeof data === 'string') {
        return data;
      }

      return data.todayEloDiff;
    },
  },
  {
    key: 'faceit.latestMatchesTrend.simple',
    description: 'Faceit matches trend',
    handler: async (_key, state) => {
      const data = await getFaceitIntegrationData(state);
      if (!data) {
        return 'Cannot fetch faceit data.';
      }
      if (typeof data === 'string') {
        return data;
      }

      return data.latestMatchesTrend.simple;
    },
  },
  {
    key: 'faceit.latestMatchesTrend.extended',
    description: 'Faceit matches trend with elo diff',
    handler: async (_key, state) => {
      const data = await getFaceitIntegrationData(state);
      if (!data) {
        return 'Cannot fetch faceit data.';
      }
      if (typeof data === 'string') {
        return data;
      }

      return data.latestMatchesTrend.simple;
    },
  },
];

File: apps/parserv2/internal/variablescache/faceit.go
package variablescache

import (
	"encoding/json"
	"errors"
	"net/http"
	model "tsuwari/parser/internal/models"

	"github.com/samber/lo"
)

type FaceitGame struct {
	Lvl int `json:"skill_level"`
	Elo int `json:"faceit_elo"`
}

type FaceitResponse struct {
	Games map[string]*FaceitGame `json:"games"`
}

type FaceitDbData struct {
	Game     *string `json:"game"`
	Username string  `json:"username"`
}

func (c *VariablesCacheService) GetFaceitData() (*FaceitGame, error) {
	c.locks.faceitIntegration.Lock()
	defer c.locks.faceitIntegration.Unlock()

	if c.cache.FaceitData != nil {
		return c.cache.FaceitData, nil
	}

	integrations := c.GetEnabledIntegrations()

	if integrations == nil {
		return nil, errors.New("integrations not enabled")
	}

	integration, ok := lo.Find(*integrations, func(i model.ChannelInegrationWithRelation) bool {
		return i.Integration.Service == "FACEIT"
	})

	if !ok {
		return nil, errors.New("faceit integration not enabled")
	}

	dbData := &FaceitDbData{}
	err := json.Unmarshal([]byte(integration.Data.String), &dbData)

	if err != nil {
		return nil, errors.New("failed to read your faceit config. Are you sure you are using integration right?")
	}

	var game string

	if dbData.Game == nil {
		game = "csgo"
	}

	client := &http.Client{}
	req, _ := http.NewRequest("GET", "https://open.faceit.com/data/v4/players?nickname="+"Satonteu", nil)
	req.Header.Set("Authorization", "Bearer "+integration.Integration.APIKey.String)
	res, err := client.Do(req)

	if err != nil {
		return nil, errors.New("failed to fetch data from faceit")
	}

	data := FaceitResponse{}

	err = json.NewDecoder(res.Body).Decode(&data)
	if err != nil {
		return nil, errors.New("failed to fetch data from faceit")
	}

	if data.Games[game] == nil {
		return nil, errors.New("Game " + game + " not found in faceit response.")
	}

	c.cache.FaceitData = data.Games[game]

	return data.Games[game], nil
}
